### PR TITLE
PATH MARKER, CAMERAS, TF

### PR DESCRIPTION
**Pull request by kev-the-dev**
- Rewrite of path marker perception, including 3d pose estimation using solvepnp and unittests
- some fixes in mission system discovered during testing/sim
- after discussion with @DSsoto, change frame ids/namespaces of cameras to increase sanity. This may cause some issues with existing software, but is a good change to make
-- Cameras are under /camera (/camera/front/left, /camera/front/right, /camera/down/left)
-- frame ids for camera match those in tf, now front_left_cam, front_right_cam, and down_left_cam

### CODE REVIEW

**Comment by kev-the-dev** [comment]
btw, we need a solution for unit test data. @whispercoros, what you think? [Catkin can download it for you](http://docs.ros.org/kinetic/api/catkin/html/howto/format2/downloading_test_data.html) with an http link, we could write our own scripts, git lfs, etc.

File: command/sub8_launch/launch/tf.launch
Comment (mattlangford) [inline command/sub8_launch/launch/tf.launch:3]: Any research into TF2? It does these static broadcasters way better with much less issues on playback

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Good overview comments, but you need some in the code as well

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Check if the denom is zero before doing this

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Zero check the denominator here as well, something above could be ensuring that but unless you're sure of that, it's best to check

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Can you disambiguate the randomish numbers

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: These should be params since it's almost certain that they won't be the same on a day to day and city to city basis. In the past there were sets of tuned numbers that you could use in the launch file so we could launch with pool numbers, sim numbers, or transdec numbers.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: The area will be dependent on the height from bottom. The mission should enforce a search height and the area should be computed using that.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/test_path_marker.py]: `np.degrees` would be less messy, and it's more of a correct angle right?

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Kalman filter would be a cool idea. Can you somehow quantify your confidence in the observation?

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: You should use a scaled cube so you can see the predicted direction

File: command/sub8_missions/missions/align_channel.py
Comment (mattlangford) [inline command/sub8_missions/missions/align_channel.py]: _Techincally_ pep8 calls for `is None` over `== None`

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Couldn't this all go up in the if statement above?

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: I'd imagine this is not very robust to noisy data. It might be good to get the height off bottom and compare with that as a sanity check as well

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: Parms

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/test_path_marker.py]: You can't subtract angles like this in the general case. Make sure the possible angles here are bounded in [-pi/2, pi/2] for this operation to be valid.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (mattlangford) [inline perception/sub8_perception/nodes/test_path_marker.py]: Is there weird spacing here or is it just my browser?

**Review by mattlangford — CHANGES_REQUESTED** [review]
I'm guessing by the tests that the algorithm works, nice job. Adding a kalman filter to filter the pose of the marker would be a cool addition that would be good to learn how to do.

I think you need to add comments before merging and there are a few small other things in my comments.

**Comment by sentree** [comment]
I asked @duggerd to create a continuous_testing directory on the SMB server. This will be used to store bags from the background Gazebo tests when that system is implemented. I could create a directory in that share for serving this data on our CI site, but I would like confirmation from @duggerd first to make sure he is fine with this solution.

**Comment by sentree** [comment]
How does this react when there are multiple visible path markers?

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: u smart

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: So far if the contour is correct, the 3d translation is quite good. I use the 2D angle for orientation, but x,y,z have been quite accurate based on bags

**Comment by kev-the-dev** [comment]
Currently, given a frame with multiple detected markers it would just return one of them at random. In my next iteration when I introduce a kalman filter I will allow for the case of two markers.

**Comment by mattlangford** [comment]
Is multiple markers something we will have to deal with?

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: The key there being a correct contour. Using competition bags?

**Comment by kev-the-dev** [comment]
I have seen one bag from competition where two markers are visible in a frame (I'll post a screenshot if I can find which one). Sounds like something AUVSI might do, so the perception should handle it.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: Yes. Every competition bag I found showing path markers completely in frame for more than a second are included in the unit tests and this node passes the 2D test. I don't know the actual 3D pose to test to but the numbers seem reasonable (Certainly correct sign for x,y,z translation and in the ballpark)

**Comment by kev-the-dev** [comment]
I believe all comments have been addressed -> trigger rereview

File: command/sub8_launch/launch/subsystems/blueview.launch
Comment (DSsoto) [inline command/sub8_launch/launch/subsystems/blueview.launch]: Would it be better if this was the dns name in case the ip changes? @whispercoros thoughts?

File: command/sub8_launch/launch/subsystems/cameras.launch
Comment (DSsoto) [inline command/sub8_launch/launch/subsystems/cameras.launch:3]: You saw both front cameras run fine with the new launch?

File: command/sub8_launch/launch/tf.launch
Comment (DSsoto) [reply command/sub8_launch/launch/tf.launch:3]: I agree here tf2, is much nicer, the migration is pretty easy to execute too.
You switch from tf to tf2_ros and instantiate a tf2_ros.TransformListener() like this 
> `listener = tf2_ros.TransformListener(tf2_ros.Buffer())`.

This line would become: 
> `<node pkg="tf2_ros" type="static_transform_publisher" name="broadcater_name"
       args="0 0 0 0 0 0 1 parent_frame child_frame" />`

Notice that there is no frequency or period argument because it is a latched publisher.

File: command/sub8_missions/missions/align_channel.py
Comment (DSsoto) [inline command/sub8_missions/missions/align_channel.py]: Can you name this something like path alignment, or something related to alignment?
Just marker feels too ambiguous.

File: command/sub8_missions/sub8/sub_singleton.py
Comment (DSsoto) [inline command/sub8_missions/sub8/sub_singleton.py]: Whats the reason for this? Doesn't this just make the code less robust  to namespace changes?

For example, we run another small sub using the same stack and push all of these topic names down under the robot nam. I think this is the purpose of using relative names for topics.

File: perception/sub8_perception/CMakeLists.txt
Comment (DSsoto) [inline perception/sub8_perception/CMakeLists.txt]: Can you put this test cmake code in a CMakeLists.txt in the test dir and just include it here using [`add_subdirectory(<path_to_testing_cmakelists>)`](https://www.johnlamp.net/cmake-tutorial-4-libraries-and-subdirectories.html#section-TestingForReal)?

This is likely to get a lot larger and I'd like to be able to quickly skim thorough the programs that are getting compiled here in this CMakeLists.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (DSsoto) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: This seems to work great so it's fine, but in the future, and for more complex shapes, considered using [`cv2.matchShapes()`](http://docs.opencv.org/3.1.0/d5/d45/tutorial_py_contours_more_functions.html) which uses the Hu moment Invariants and should be much more robust than ad-hoc matching heuristics you come up with.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (DSsoto) [inline perception/sub8_perception/nodes/follow_orange_pipes.py]: This should be a ros param. You could also replace the `marker.scale.*` magic constants in the rviz pub function with an expression using it.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (DSsoto) [inline perception/sub8_perception/nodes/test_path_marker.py]: This file should go in the test directory.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (DSsoto) [inline perception/sub8_perception/nodes/test_path_marker.py]: lol if it fails it's bullshit! +1

File: utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py
Comment (DSsoto) [inline utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py]: Good stuff!

**Review by DSsoto — CHANGES_REQUESTED** [review]
Requesting a couple of changes and have a couple of comments I'd like addressed. Would like to merge today if possible.

File: command/sub8_launch/launch/tf.launch
Comment (RustyBamboo) [inline command/sub8_launch/launch/tf.launch]: I think the blueview orientation is <0 0 -0.707 0.707> relative to base_link, where the x-axis is pointing right, y-axis is forward, and z-axis is up.

Comment (mattlangford) [reply command/sub8_launch/launch/tf.launch]: Is that the case? I thought the sonar was pointing straight forward?

File: command/sub8_missions/missions/align_channel.py
Comment (mattlangford) [reply command/sub8_missions/missions/align_channel.py]: Yeah, I didn't like all the mixed references to marker/orange_path/alignment/etc last year

Also you should use fprint for the colors and headers (since there's a header in every print line)

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: Does it need to be? It won't change will it? It's better not to clutter up the things that we could change with things that never would.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (mattlangford) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: That brings up a good point of non-fully in frame markers, do you handle that? And only on 2 transdec bags?

File: utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py
Comment (mattlangford) [inline utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py]: I see your two spaces, don't think I've gone blind in my old age.

File: utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py
Comment (mattlangford) [inline utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py]: This could be a useful general tool, maybe add an option to specify `from` and `to` topics where you can call it like: `fix_bag.py --in dumb_topic_name --out good_topic_name --in another_bad_topic --out good_topic_2`. That would fix topics but you'd have to specify framing in a similar way. This would complicate the thing this file is doing right now but make it a lot more general.

**Review by mattlangford — APPROVED** [review]
I'm cool with this. I think any changes on the base algorithm (kalman filter, more robust detection, knowing if the thing isn't fully in frame) could go in another PR. Defering to David for final verdict.

I would like to see fprint used in the mission.

File: command/sub8_launch/launch/tf.launch
Comment (kev-the-dev) [reply command/sub8_launch/launch/tf.launch]: It is, but the data it produces treats X, Y, and Z as Daniel described. So having the correct transform here would be nice so the ogrid/pointcloud published from the data will transform correctly.

File: command/sub8_launch/launch/tf.launch
Comment (kev-the-dev) [reply command/sub8_launch/launch/tf.launch:3]: Let's not switch to tf2 in the PR. Seems like it might offer some benefits, but not sure what it will break. Specifically, I don't know if txros supports it, which could break missions.

File: command/sub8_missions/sub8/sub_singleton.py
Comment (kev-the-dev) [reply command/sub8_missions/sub8/sub_singleton.py]: I changed it because I was having missions get stuck waiting for odom forever in sim, doesn't seem to be happening any more... huh. I'll switch it back

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: The notes.txt for the transdec bags says only 3 have path markers, and one was half out of frame :(. If it's out of frame, I don't detect it

File: command/sub8_launch/launch/subsystems/blueview.launch
Comment (kev-the-dev) [reply command/sub8_launch/launch/subsystems/blueview.launch]: Don't know the domain and not sure if the sdk handles domains, but I'm happy to try it.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/test_path_marker.py]: Okay, I moved it. It might cause issues though because it is not in one of the two default folders for python nodes. Do I need to update setup.py?

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (kev-the-dev) [reply perception/sub8_perception/nodes/test_path_marker.py]: It's so glarry I can barely identify it myself. Hope to get a few more testing bags once new down camera is installed.

File: utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py
Comment (kev-the-dev) [reply utils/sub8_ros_tools/sub8_ros_tools/fix_bag.py]: Good idea, I made it a generic tool and added a [wiki page](https://github.com/uf-mil/SubjuGator/wiki/Working-with-old-bags) for using it with subjugator bags. I'll probably move it to mil_common once it has a tools package in master branch.

File: perception/sub8_perception/nodes/test_path_marker.py
Comment (DSsoto) [reply perception/sub8_perception/nodes/test_path_marker.py]: Nothing should really need to import from this right? And yea, if anything breaks, add an __init__.py and update the setup.py.

File: perception/sub8_perception/nodes/follow_orange_pipes.py
Comment (DSsoto) [reply perception/sub8_perception/nodes/follow_orange_pipes.py]: I kind of like the idea of each vision node having a yaml where all it's relevant parameters are defined. That way if anything needs to be changed, everyone knows where to go. No need to skim through all the source code.

I'll leave it up to Kevin's discretion.

File: command/sub8_launch/launch/tf.launch
Comment (DSsoto) [reply command/sub8_launch/launch/tf.launch:3]: Sure, but after this, I will be very hesitant to merge stuff not using tf2. Tf has been deprecated since late 2013 (Hydro) and is pretty buggy.

**Review by DSsoto — APPROVED** [review]

File: command/sub8_launch/launch/subsystems/blueview.launch
Comment (DSsoto) [reply command/sub8_launch/launch/subsystems/blueview.launch]: mil-com-teledyne-p900.ad.mil.ufl.edu

File: command/sub8_launch/launch/subsystems/cameras.launch
Comment (kev-the-dev) [reply command/sub8_launch/launch/subsystems/cameras.launch:3]: Yes. Both looked good besides green bug and had correct frames in camera info and image topics

Comment (kev-the-dev) [reply command/sub8_launch/launch/subsystems/cameras.launch:3]: Also going to disable down camera in launch because it is not connected.

**Comment by kev-the-dev** [comment]
Updated to use DNS for blueview. Good to go!

**Comment by mattlangford** [comment]
Someone should try tf2 with txros, and make a PR to support it if it doesn't work

**Pull request merged**
